File: src/main/java/net/sf/rails/game/state/Change.java
package net.sf.rails.game.state;
/**
 * Base Class for all Change Objects
 * 
 * Replaces previous move interface
 */
public abstract class Change {
    
    protected void init(State state){
        state.getStateManager().getChangeStack().addChange(this);
    }

    abstract void execute();
    abstract void undo();  
    public abstract State getState();

}

File: dotnet/RailsLib.Net/Net/Game/State/Change.cs
using System;

/**
 * Base Class for all Change Objects
 * 
 * Replaces previous move interface
 */
namespace GameLib.Net.Game.State
{
    abstract public class Change 
    {
        protected void Init(GameState state)
        {
            state.StateManager.ChangeStack.AddChange(this);
        }

        public abstract void Execute();
        public abstract void Undo();
        public abstract GameState GameState { get; }
        public T GetGameState<T>() where T : GameState
        {
            return (T)GameState;
        }
    }
}
